### ISSUE TITLE
Research approaches of making certain fields (statuses) computable based on other fields

### ISSUE THREAD

**Issue by tskir** [EBIvariation/trait-curation#41]
Related to this discussion https://github.com/EBIvariation/trait-curation/pull/35#discussion_r440162367

**Comment by tskir** [comment]
@joj0s In addition to what you mentioned:
* Keep it as a CharField and add a check whenever a review is issued or whenever a mapping is changed to automatically compute the field's value
* Use this package https://pypi.org/project/django-computedfields/

I also found this package: https://github.com/schinckel/django-computed-field.

This is indeed a tough choice, with many things to consider.
* If we have to re-compute the fields when certain events happen:
  + (+) It will be more efficient and rely only on core Django functionality
  + (–) It's more complicated and leaves possibility for race conditions and inconsistent database states
* If we use external libraries:
  + (+) It's much easier to implement
  + (–) It will be much less efficient (since presumably the fields will be recomputed on each query to them). Also, both of the packages listed above have tiny userbase (judging by the number of GitHub stars/forks), so we don't have any guarantee they will continue to work in the future indefinitely, as Django/Python versions update.

Let's leave this an open question for now. If you have any additional thoughts, please post them here.

**Comment by jerch** [comment]
@tskir Maintainer of https://pypi.org/project/django-computedfields/ here. The package is meant as a denormalization pattern by moving the calculation burden from select queries to inserts/updates, thus is not a good answer to "any calculation purpose".

May I ask what your field computations look like and what you want to optimize? 

> It will be much less efficient (since presumably the fields will be recomputed on each query to them)

Thats not true for django-computedfields, it only triggers a field update on dependency changes down to record level. Thats the reason, why it is needed to place a dependency listing on a computed field (which admittedly can get exhaustive).

> Also, both of the packages listed above have tiny userbase (judging by the number of GitHub stars/forks), so we don't have any guarantee they will continue to work in the future indefinitely, as Django/Python versions update.

Thats true. Also note that the package is still at beta stage, and the package is about to get a major refactoring before going into more silent waters (removing custom metaclass). So not sure if you are up for beta testing. A stable release will take about 2 more months or so.
The code itself only relies on ORM semantics, no hidden DBMS magic with custom SQL as with other packages. Unless Django turns everything upside-down, there is a high chance it keeps working for several versions (this worked so far from 1.11 up to 3.0).

**Comment by tskir** [comment]
Hi @jerch! Thank you for coming to this thread & commenting, this was unexpected for us, but very useful :)

> > It will be much less efficient (since presumably the fields will be recomputed on each query to them)
>
> Thats not true for django-computedfields, it only triggers a field update on dependency changes down to record level.

This is great and sounds exactly like something we would want. We have some fields which depend on other fields, and we'd like to have a way to seamlessly update them, but do not want to recompute them every time they are used.

> Thats the reason, why it is needed to place a dependency listing on a computed field (which admittedly can get exhaustive).

I think that's fine. It's better to have an explicit list of dependencies rather than relying on “magical” auto-update.

Our use cases are rather simple. For example:
* Field A can be "reviewed" or "not_reviewed"
* Field B can contain some values like "current", "deleted" and some more (not important)
* We want field C to be computed like this:
  + If A is "not_reviewed", then C should be also "not_reviewed"
  + If A is "reviewed", then C should mirror the value of B

Looking at documentation, it looks like we could implement it using your module. Please correct me if I'm wrong.

Finally, regarding the beta status of the module. Since it looks like it's being very actively developed, I think we would be willing to at least try it, participate in beta testing, and report to you if we find any problems.

**Comment by jerch** [comment]
Yes your use case would work with the package. Something like this should do the trick:
```python
class SomeModel(ComputedFieldsModel):
    fieldA = CharField(...)
    fieldB = CharField(...)
    
    @computed(CharField(...), depends=[['self', ['fieldA', 'fieldB']]])
    def fieldC(self):
        return self.fieldB if self.fieldA == 'reviewed' else self.fieldA
```

But note - thats a typical case, where an annotation field should be preferred instead (unless query perf suffers to much):
```python
SomeModel.objects.annotate(fieldC=Coalesce(
    NullIf('fieldA', Value('reviewed')),    # pull fieldA if its not set to 'reviewed'
    'fieldB',                               # else pull fieldB
    output_field=CharField(...)
))
```

**Comment by tskir** [comment]
@jerch Thank you for a comprehensive overview of all of the different methods we could use, it's really appreciated! While for now we only have simple use cases in mind, and `Coalesce()` would probably suffice, it's quite likely that in the future we will have more complicated scenarios, and in this case we'll definitely consider to give django-computedfields a go.

**Comment by jerch** [comment]
> ... While for now we only have simple use cases in mind, and Coalesce() would probably suffice ...

Yes good decision, let the DBMS do the work it was made for. :smile_cat:

**Comment by tskir** [comment]
@joj0s Additional point: research whether annotated fields work on push or on pull

**Comment by joj0s** [comment]
@tskir So I have been researching the above, here's what I've found so far:

1. We cannot use `Coalesce()` to annotate our fields because we do not have a `CharField` containing a 'is_reviewed' or 'not_reviewed' value, but it is rather a `is_reviewed` boolean field. So we can't just pull its value. I am researching an approach on using [conditional expressions](https://docs.djangoproject.com/en/3.0/ref/models/conditional-expressions/) to achieve what we want, should we choose the annotated fields route.

2. The `annotate` function is used to add fields to queries, so we can't declare one in the model specification like the rest of the regular fields, nor can we create a custom `create` method for a model instance, since it cannot be used on a single object, only on query sets. So I am not sure if there is a good way to instantiate them.

**Comment by tskir** [comment]
> Alternatively we could replace the BooleanField with a CharField containing actual Status values.

I don't really want to do that, I mean boolean data type exists for a very good reason :)

Since the simpler approaches cause pain, let's just go ahead with the django-computedfields package in the way described here: https://github.com/EBIvariation/trait-curation/issues/41#issuecomment-649089053

**Comment by joj0s** [comment]
I can confirm everything works great with the package. Once we get a larger number of traits to process, I will look into query performance as well.

**Comment by tskir** [comment]
Since research is complete, I am closing this issue. Computable fields will first be implemented in #25.
